src/engine/source/router/src/runtimePolicy.cpp

```diff
@@ -32,14 +32,142 @@ std::optional<base::Error> RuntimePolicy::build(std::shared_ptr<builder::Builder
+            {
+                const std::string& key = matchVerbose[1].str();
+                std::shared_ptr<std::stringstream> traceStream = std::make_shared<std::stringstream>();
+                *traceStream << trace;
+
+                // Check if an entry with the same key already exists in the first level of m_traceBuffer
+                auto outerIt = m_traceBuffer.find(m_asset);
+                if (outerIt == m_traceBuffer.end())
+                {
+                    // There is no entry for m_asset, create a new one
+                    std::unordered_map<std::string, std::vector<std::shared_ptr<std::stringstream>>> innerMap;
+                    innerMap[key] = std::vector<std::shared_ptr<std::stringstream>> {traceStream};
+
+                    m_traceBuffer[m_asset] = innerMap;
+                }
+                else
+                {
+                    // Check if an entry with the same key already exists in the second level of m_traceBuffer
+                    auto& innerMap = outerIt->second;
+                    auto innerIt = innerMap.find(key);
+                    if (innerIt == innerMap.end())
+                    {
+                        // There is no entry for the key in the second level, create a new one
+                        innerMap[key] = std::vector<std::shared_ptr<std::stringstream>> {traceStream};
+                    }
+                    else
+                    {
+                        // The key already exists at the second level, add the traceStream to the existing list
+                        innerIt->second.push_back(traceStream);
+                    }
+                }
+            }
+        }));
+}
+
+const std::variant<std::tuple<std::string, std::string>, base::Error>
+RuntimePolicy::getData(const std::string& policyName, DebugMode debugMode)
+{
+    std::unique_lock<std::shared_mutex> lock {m_mutexData};
+    auto trace = json::Json {R"({})"};
+    if (DebugMode::OUTPUT_AND_TRACES_WITH_DETAILS == debugMode)
+    {
+        if (m_history[policyName].empty())
+        {
+            return base::Error {fmt::format(
+                "Policy '{}' has not been configured for trace tracking and output subscription", policyName)};
+        }
+
+        for (const auto& [asset, condition] : m_history[policyName])
+        {
+            if (m_traceBuffer.find(policyName) == m_traceBuffer.end())
+            {
+                return base::Error {fmt::format(
+                    "Policy '{}' has not been configured for trace tracking and output subscription", policyName)};
+            }
+            auto& tracePair = m_traceBuffer[policyName];
+            if (tracePair.find(asset) != tracePair.end())
+            {
+                auto& traceVector = tracePair[asset];
+                std::set<std::string> uniqueTraces; // Set for warehouses single traces
+                for (const auto& traceStream : traceVector)
```

Comment:
this could be merge like this, 
`for (const auto& traceStream : tracePair[asset])`
right?

Reply:
It could be indeed, we do find this one arrangement clarifying and it does not have any considerable impact